sim/test/src/zynq/tcu/AXIWrapperTCUSpec.scala

```diff
@@ -497,6 +499,89 @@ class AXIWrapperTCUSpec extends FunUnitSpec {
               }
             }
 
+          def sample(
+              programSize: Int,
+              interval: Int,
+              blockSize: Int
+          ) =
+            it(
+              s"should sample with size=$programSize, interval=$interval, blockSize=$blockSize",
+              Slow
+            ) {
+              test(
+                new AXIWrapperTCU(
+                  gen,
+                  layout,
+                  AXIWrapperTCUOptions(
+                    inner = TCUOptions(sampleBlockSize = blockSize),
+                    dramAxiConfig = axiConfig
+                  )
+                )
+              ).withAnnotations(Seq(VerilatorBackendAnnotation)) { m =>
+                m.setClocks()
+                m.clock.setTimeout(Int.MaxValue)
+
+                implicit val layout: InstructionLayout =
+                  m.setInstructionParameters()
+
+                val cycleCount = 3 * programSize
+
+                // drams to listen
+                fork {
+                  dram0.listen(m.clock, m.dram0)
+                }
+                fork {
+                  dram1.listen(m.clock, m.dram1)
+                }
+                fork {
+                  m.sample.ready.poke(true.B)
+
+                  val samples = (0 until cycleCount / interval).map({
+                    case 0 => (1, false)
+                    case x =>
+                      (
+                        Math.min(x * interval - 1, 1000),
+                        (x + 1) % blockSize == 0
+                      )
+                  })
+
+                  for ((programCounter, last) <- samples) {
+                    m.sample.waitForValid()
+                    m.sample.bits.bits.programCounter.expect(programCounter.U)
+                    m.sample.bits.last.expect(last.B)
+
+                    m.clock.step()
+                  }
+                }
+
+                fork {
+                  var pc = 0
+
+                  m.instruction.enqueue(
+                    Instruction(
+                      Opcode.Configure,
+                      ConfigureArgs(Configure.sampleInterval, interval)
+                    )
+                  )
+
+                  m.instruction.enqueue(
+                    Instruction(
+                      Opcode.Configure,
+                      ConfigureArgs(Configure.programCounter, 0)
+                    )
+                  )
+
+                  for (_ <- 0 until programSize) {
+                    m.instruction.enqueue(Instruction(Opcode.NoOp))
+                  }
+                }
+
+                for (_ <- 0 until cycleCount) {
```

Comment:
For the future: you can avoid this cycle count kludge by keeping a reference to the threads launched and then joining them all as the last step in the test. See the TCU tests for some examples.